### PR TITLE
[tool] Add a way to opt a file out of formatting

### PR DESCRIPTION
**Pull request by Hixie**
Per discussion on Discord, this adds a way to opt-out of the formatting.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by Hixie** [comment]
I made this apply to all files rather than just Dart, mostly because it was easier to plug into the code where `File` objects were still around. I considered pushing the File-to-String conversion lower down the codepaths, but that ended up being pretty invasive and not obviously better. I also considered just having the Dart code go back from String to Files but that ended up being pretty ugly too.

File: script/tool/lib/src/format_command.dart
Comment (stuartmorgan) [inline script/tool/lib/src/format_command.dart:212]: > I made this apply to all files rather than just Dart, mostly because it was easier to plug into the code where `File` objects were still around.

I would much rather this apply only to Dart, as this is strictly worse than `clang-format off` for `clang-format`-formatted files because the latter will prevent people with editors that are configured to auto-format from accidentally formatting the file when editing it. (I really, really wish `dartfmt` would support this for the same reason.)

What's the issue with just adding a `file.basename.endsWith('.dart')` check inside this lambda?

File: script/tool/CHANGELOG.md
Comment (stuartmorgan) [inline script/tool/CHANGELOG.md:14]: The period needs to be inside the \`s; it's part of the string being searched for.

File: script/tool/lib/src/format_command.dart
Comment (stuartmorgan) [inline script/tool/lib/src/format_command.dart:209]: I think it's worth adding a short explanatory comment here about why this exists to save someone finding the PR to understand the reason.

File: script/tool/test/format_command_test.dart
Comment (stuartmorgan) [inline script/tool/test/format_command_test.dart:102]: This is a distinct enough thing from the basic "does the formatter get called" that I think it would be much better to have a separate "skips files with the opt-out" test, so that a test failure would be more specific.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: script/tool/lib/src/format_command.dart
Comment (Hixie) [reply script/tool/lib/src/format_command.dart:220]: please update the style guide to mention this

File: script/tool/lib/src/format_command.dart
Comment (Hixie) [reply script/tool/lib/src/format_command.dart:212]: it seemed weird to have this code know about file extensions too, but i'm happy to add it if you like

File: script/tool/lib/src/format_command.dart
Comment (Hixie) [reply script/tool/lib/src/format_command.dart:220]: since i added the check for the extension, i was able to remove all this error handling logic and thus the comment is gone too

**Comment by Hixie** [comment]
PTAL

File: script/tool/lib/src/format_command.dart
Comment (stuartmorgan) [reply script/tool/lib/src/format_command.dart]: Do you still want me to add this to the style guide? I'm happy to do so, but it's also a position that not everyone agrees with.

File: script/tool/CHANGELOG.md
Comment (stuartmorgan) [inline script/tool/CHANGELOG.md:16]: This got merged into the wrong section, it needs to be in a new one since 0.6.0 is published.

You should actually do a version bump in this PR since in order to use it in flutter/packages it'll need to be published (and then the flutter/packages pin rolled to pick it up)

File: script/tool/lib/src/format_command.dart
Comment (stuartmorgan) [inline script/tool/lib/src/format_command.dart:213]: 🤣 Well, I did ask.

(I was just thinking something like "Allow package maintainers to opt out of Dart autoformatting in order to use Flutter's manual formatting style instead.", but this certainly works.)

**Review by stuartmorgan — APPROVED** [review]
LGTM modulo adding a version change.

File: script/tool/CHANGELOG.md
Comment (stuartmorgan) [reply script/tool/CHANGELOG.md:16]: Oh, and it should be changed to "skips Dart files".

File: script/tool/lib/src/format_command.dart
Comment (Hixie) [reply script/tool/lib/src/format_command.dart]: I don't feel strongly about it (but then I get a guilty pleasure in writing sentences that use the same pronoun to mean different people), but if there's an argument to be made that it makes things clearer, we should document it somewhere people can see it.

File: script/tool/CHANGELOG.md
Comment (Hixie) [reply script/tool/CHANGELOG.md:16]: Done.

**Pull request merged**
